### PR TITLE
Fix network simplex callable

### PR DESCRIPTION
**Pull request by ShreySharma07**
## Description
This PR updates `network_simplex` to support callable objects for the `capacity` parameter, aligning it with other flow algorithms like `edmonds_karp` and `preflow_push`.

Previously, `_DataEssentialsAndFunctions` assumed `capacity` was always a string attribute key. If a callable was passed, it would default to infinite capacity without warning. This change adds logic to check `callable(capacity)` and extract values correctly.

## Related Issue
Closes #8406 (Reopened item regarding missing algorithms).

## Verification
- Validated that `network_simplex` now correctly executes the callable to determine capacity.
- Ran existing tests `networkx/algorithms/flow/tests/test_networksimplex.py` (All 31 passed).

### CODE REVIEW

**Review by dschult — COMMENTED** [review]
The test should be rewritten to use pytest and put into test_networksimplex. I suggest looking at that file and copy/changing an existing test so the mechanics are similar but it tests what you want it to test.

It also looks like the .gitignore file has been changed so that the end of file character is not there.

The code itself is much more complicated than it needs to be. Take a look at the way this is handled in other files for the weight_function.  And the point is to skip edges where the function returns `None`, not where the edge value is zero.

File: networkx/algorithms/flow/networksimplex.py [hunk 1]
```diff
@@ -36,15 +36,43 @@ def __init__(
             edges = G.edges(data=True, keys=True)
 
         inf = float("inf")
-        edges = (e for e in edges if e[0] != e[1] and e[-1].get(capacity, inf) != 0)
```
Comment (amcandio) [inline networkx/algorithms/flow/networksimplex.py:39]: You can go back to this format if you use the walrus operator

**Comment by ShreySharma07** [comment]
> The test should be rewritten to use pytest and put into test_networksimplex. I suggest looking at that file and copy/changing an existing test so the mechanics are similar but it tests what you want it to test.
> 
> It also looks like the .gitignore file has been changed so that the end of file character is not there.
> 
> The code itself is much more complicated than it needs to be. Take a look at the way this is handled in other files for the weight_function. And the point is to skip edges where the function returns `None`, not where the edge value is zero.

Thanks for the review! I've updated the PR with the following changes:

Tests: I moved the new tests into networkx/algorithms/flow/tests/test_networksimplex.py and converted them to the standard pytest format as requested.

Gitignore: I reverted the accidental changes to .gitignore.

Logic: I refactored the capacity extraction to be cleaner.

Note on Zero-Capacity Edges: regarding the suggestion to keep edges where capacity=0: I found that passing zero-capacity edges to the solver caused it to hang (infinite loop) on test_zero_capacity_edges. To prevent this regression, I maintained the filter to skip edges where capacity is 0 or None. I also verified this behaviour with a new test case.

All CI checks are passing now.

File: networkx/algorithms/flow/networksimplex.py (hunk 1 above)
Comment (ShreySharma07) [reply networkx/algorithms/flow/networksimplex.py:39]: Thanks for the guidance! I've updated the PR to address the feedback:

Refactoring: @amcandio I implemented your suggestion to use the walrus operator (:=), allowing me to refactor the edge filtering into a single, clean generator expression as requested.

Tests & Logic: @dschult

I moved the new tests into networkx/algorithms/flow/tests/test_networksimplex.py using the standard pytest format.

I verified that the .gitignore file is back to its original state.

Note on Zero-Capacity Edges: I retained the check cap != 0. I found that if I removed it (allowing 0-capacity edges into the solver), the algorithm would hang (infinite loop) during regression tests like test_zero_capacity_edges. The current logic prevents this hang while correctly handling the None (hidden) edges.

File: networkx/algorithms/flow/networksimplex.py [hunk 2]
```diff
@@ -36,14 +36,35 @@ def __init__(
...
+
+        # Getting capacity (String vs Function)
+        if callable(capacity):
+
+            def get_cap(e):
+                return capacity(e[0], e[1], e[-1])
+        else:
+
```
Comment (amcandio) [inline networkx/algorithms/flow/networksimplex.py:41]: Nit: inf should be declared here

File: networkx/algorithms/flow/networksimplex.py [hunk 3]
```diff
@@ -36,14 +36,35 @@ def __init__(
...
         for i, e in enumerate(edges):
             self.edge_sources.append(self.node_indices[e[0]])
             self.edge_targets.append(self.node_indices[e[1]])
+
             if multigraph:
                 self.edge_keys.append(e[2])
+
             self.edge_indices[e[:-1]] = i
-            self.edge_capacities.append(e[-1].get(capacity, inf))
+
```
Comment (amcandio) [inline networkx/algorithms/flow/networksimplex.py:65]: Nit: I'd remove this whitespace changes

File: networkx/algorithms/flow/tests/test_networksimplex.py [hunk 4]
```diff
@@ -479,3 +479,40 @@ def test_network_simplex_unbounded_flow():
...
+    assert min_cost == 5  # 5 units * 1 cost
+    assert flow_dict["s"]["t"] == 5
+    assert flow_dict["t"].get("s", 0) == 0
+
+
+def test_callable_capacity_none():
+    """Test that callable returning None effectively removes the edge."""
+    G = nx.DiGraph()
+    G.add_edge("s", "t", count=10, cost=1)
+    G.add_edge("s", "hidden", count=10, cost=1)
```
Comment (amcandio) [inline networkx/algorithms/flow/tests/test_networksimplex.py:491]: You should connect hidden to t so the test actually filters the edge out

Comment (ShreySharma07) [reply networkx/algorithms/flow/tests/test_networksimplex.py:491]: Thanks @amcandio  I've connected hidden to t to ensure the edge is skipped solely due to the None return. I also moved inf into the else block and cleaned up the whitespace.

File: networkx/algorithms/flow/tests/test_networksimplex.py [hunk 5]
```diff
@@ -479,3 +479,25 @@ def test_network_simplex_unbounded_flow():
         match="negative cycle with infinite capacity found",
     ):
         nx.network_simplex(G)
+
+
+def test_callable_capacity_none():
+    """Test that callable returning None effectively removes the edge."""
+    G = nx.DiGraph()
+    G.add_edge("s", "t", count=10, cost=1)
```
Comment (amcandio) [inline networkx/algorithms/flow/tests/test_networksimplex.py]: This is still the shortest so this test is not really exercising the filtering. One way to see if a test is good is locally removing part of your logic to see if the test would catch that bug

Comment (ShreySharma07) [reply networkx/algorithms/flow/tests/test_networksimplex.py]: @amcandio 
I've updated the test to construct a cheaper path through the hidden node. Now the test asserts that the solver is forced to take the expensive direct path, which proves the None filtering is actively working.

**Review by amcandio — DISMISSED** [review]
LGTM

**Comment by gulshan-123** [comment]
I can see some lines where it is assumed capacity is string. for example see L 637, L 680, etc. Does this need not to be changed?

Also, functions in mincost.py directly uses network_simplex. Should we change docs there too?
And since network_simplex also have parameters weight, and demand, should we add support for them as callables too (maybe in separate PR)?

**Comment by amcandio** [comment]
> I can see some lines where it is assumed capacity is string. for example see L 637, L 680, etc. Does this need not to be changed?
> 
> Also, functions in mincost.py directly uses network_simplex. Should we change docs there too?
> And since network_simplex also have parameters weight, and demand, should we add support for them as callables too (maybe in separate PR)?

Good catch!

**Comment by ShreySharma07** [comment]
> I can see some lines where it is assumed capacity is string. for example see L 637, L 680, etc. Does this need not to be changed?
> 
> Also, functions in mincost.py directly uses network_simplex. Should we change docs there too? And since network_simplex also have parameters weight, and demand, should we add support for them as callables too (maybe in separate PR)?

thanks man @gulshan-123  to look in this I will update it in the pr.

File: networkx/algorithms/flow/networksimplex.py [hunk 6]
```diff
@@ -35,36 +35,37 @@ def __init__(
...
-        edges = (
-            e
-            for e in edges
-            if e[0] != e[1] and (cap := get_cap(e)) is not None and cap != 0
-        )
+            # Calculate Capacity
+            if callable(capacity):
+                cap = capacity(u, v, d)
+            else:
+                cap = d.get(capacity, inf)
```
Comment (dschult) [inline networkx/algorithms/flow/networksimplex.py]: I think we should handle the `capacity` argument like we handle the weight argument for shortest_paths:
- turn the input into a callable if it isn't already.
- Implement that process as a decorator and apply the decorator to functions that have a capacity parameter.

This will remove the need for this ugly code that checks `callable(capacity)` for every single edge.

Is this a reasonable approach?  Is there something holding us back from doing this?

Comment (ShreySharma07) [reply networkx/algorithms/flow/networksimplex.py]: @dschult Thanks for the pointer! I've implemented the normalization pattern similar to `_weight_function` in `shortest_paths/weighted.py`.

Helper Function: I added `_capacity_callable(capacity)` which standardizes the input into a function at the start of network_simplex.

Cleanup: I removed all the repetitive if callable(capacity): checks inside the inner loops of `_DataEssentialsAndFunctions` and the main function. The loops now simply call` capacity(u, v, d)`.

Writing a robust decorator that handles *args and **kwargs correctly for a function with complex defaults (like network_simplex) is error-prone. The manual call is fail-safe.
I chose this explicit conversion over a custom decorator to match the existing pattern in `shortest_paths/weighted.py` (see `_weight_function`). This seemed clearer and less invasive than adding a new decorator implementation for a single function usage.

This should be much cleaner and more performant. Pushed and ready for review!

Comment (gulshan-123) [reply networkx/algorithms/flow/networksimplex.py]: I think _capacity_function is already implemented in flow/utils.py, you should import that.

Also, just a suggestion, try to not modify list comprehension instead of making it in expanded form. You may look other files of `flow` dir to see how that is done using `:=` operator, etc, if you are comfortable using that.
Thank you!

Comment (ShreySharma07) [reply networkx/algorithms/flow/networksimplex.py]: @gulshan-123 I missed that _capacity_function was already available in `utils.py`.

I've updated the code to import and use `networkx.algorithms.flow.utils._capacity_function` instead of redefining it.

I reverted the explicit loop in `__init__` back to a generator expression using the walrus operator `:=`, while keeping the logic correct.

Thanks for guiding me to the existing utility!

**Review by rossbar — COMMENTED** [review]
This has been static for a while - are you still working on it @ShreySharma07 ? It seems to me there are quite a few unrelated changes that make this more difficult to review. I'd recommend backing out every change that is not directly related to the proposed feature in order to help move this along.
